Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Runtime/InitializeAddressables.cs

```diff
@@ -0,0 +1,118 @@
+                string[] coreUnityAssetPackNames = AndroidAssetPacks.GetCoreUnityAssetPackNames(); // only returns names of asset packs that are fast-follow or on-demand delivery
+                if (coreUnityAssetPackNames.Length == 0)
+                    Debug.LogError("Cannot retrieve core Unity asset pack names. PlayCore Plugin is not installed.");
+                else
+                    AndroidAssetPacks.DownloadAssetPackAsync(coreUnityAssetPackNames, CheckDownloadStatus);
+            }
+#else
+            Setup();
+#endif
+        }
+
+        void Setup()
+        {
+            Addressables.ResourceManager.ResourceProviders.Add(new PlayAssetDeliveryAssetBundleProvider());
+            Addressables.ResourceManager.InternalIdTransformFunc = AssetPackTransformFunc;
+            hasInitialized = true;
+        }
+
+        void CheckDownloadStatus(AndroidAssetPackInfo info)
+        {
+            if (info.status == AndroidAssetPackStatus.Failed)
+                Debug.LogError($"Failed to retrieve the state of asset pack '{info.name}'."); 
+            else if (info.status == AndroidAssetPackStatus.Unknown)
+                Debug.LogError($"Asset pack '{info.name}' is unavailable for this application. This can occur if the app was not installed through Google Play.");
+            else if (info.status == AndroidAssetPackStatus.Canceled)
+                Debug.LogError($"Cancelled asset pack download request '{info.name}'.");
+            else if (info.status == AndroidAssetPackStatus.WaitingForWifi)
+            {
+                AndroidAssetPacks.RequestToUseMobileDataAsync(result =>
+                {
+                    if (!result.allowed)
+                        Debug.LogError("Request to use mobile data was denied.");
+                });
+            }
+            else if(info.status == AndroidAssetPackStatus.Completed)
+            {
+                string assetPackPath = AndroidAssetPacks.GetAssetPackPath(info.name);
+                if (string.IsNullOrEmpty(assetPackPath))
+                {
+                    Debug.LogError($"Downloaded asset pack '{info.name}' but cannot locate it on device.");
+                }
+            }
+        }
+    
+        string AssetPackTransformFunc(IResourceLocation location)
+        {
+
+#if UNITY_ANDROID && !UNITY_EDITOR
+            string filename = Path.GetFileName(location.InternalId);
+
+            if (location.ResourceType == typeof(IAssetBundleResource))
+            {
+                string assetPackName = Path.GetFileNameWithoutExtension(location.InternalId);
+                if (PlayerPrefs.HasKey(assetPackName))
+                {
+                    // Load bundle from a custom asset pack. PlayAssetDeliveryBundleProvider.Provider previously stored its path.
+                    return PlayerPrefs.GetString(assetPackName);
+                }
+            }
+
+            string streamingAssetsPackPath = AndroidAssetPacks.GetAssetPackPath("UnityStreamingAssetsPack");
```

Comment:
Users should never have to hardcode this pack name in their code. This is the first sign that something is being done in an unsupported way and might cause problems in the future.
I don't know exactly how addressables set up works, but just to make sure we understand each other - bundles created by addressables package will end up in streaming assets location (no matter where it is) only if they were dropped in project's "Assets/StreamingAssets" directory. If the bundle was added in a directory ending with .androidpack, then that bundle will end up in a separate asset pack, not in streaming assets path. Our VFS might be able to locate the asset in it's own install-time asset pack even when streaming assets path is provided, however it might not work in all cases and could stop working in the future without anyone in VFS / mobile team being aware that it would affect addressables.

Reply:
Hmm Addressables has a custom build processor (AddressablesPlayerBuildProcessor.cs) that temporarily moves the build folder (Library/com.unity.addressables) containing bundles, catalog.json, settings.json to Assets/StreamingAssets. Simply removing the bundles from the build folder (what I basically did in BuildScriptPlayAssetDelivery.cs) will prevent them from being added to the streaming assets path. 

By default Addressables expects to locate content (bundles, catalog.json, settings.json) in the Application.streamingAssetsPath directory. I'm guessing these files won't exist in Application.streamingAssetsPath if the streaming assets pack is fast-follow or on-demand. 

Is there any other way to retrieve the streaming assets pack name? AndroidAssetPacks.GetCoreUnityAssetPackNames() can return it, but not sure if there is a way to tell which pack name is the right one.